core/server/master/src/main/java/alluxio/master/file/RpcContext.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file;
+
+import alluxio.exception.status.UnavailableException;
+import alluxio.master.journal.JournalContext;
+import alluxio.master.journal.NoopJournalContext;
+
+import com.google.common.base.Throwables;
+
+import java.io.Closeable;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Context passed through the span of a file system master RPC to manage actions that need to be
+ * taken when the body of the RPC is finished.
+ */
+@NotThreadSafe
+public final class RpcContext implements Closeable {
+  public static final RpcContext NOOP =
+      new RpcContext(NoopBlockDeletionContext.INSTANCE, NoopJournalContext.INSTANCE);
+
+  @Nullable
+  private final BlockDeletionContext mBlockDeletionContext;
+  private final JournalContext mJournalContext;
+
+  // Used during close to keep track of thrown exceptions.
+  private Throwable mThrown = null;
+
+  /**
+   * Creates an {@link RpcContext}. This class aggregates different contexts used over the course of
+   * an RPC, and makes sure they are closed in the right order when the RPC is finished.
+   *
+   * @param blockDeleter block deletion context
+   * @param journalContext journal context
+   */
+  public RpcContext(BlockDeletionContext blockDeleter, JournalContext journalContext) {
+    mBlockDeletionContext = blockDeleter;
+    mJournalContext = journalContext;
+  }
+
+  /**
+   * @return the journal context
+   */
+  public JournalContext getJournalContext() {
+    return mJournalContext;
+  }
+
+  /**
+   * @return the block deletion context
+   */
+  public BlockDeletionContext getBlockDeletionContext() {
+    return mBlockDeletionContext;
+  }
+
+  @Override
+  public void close() throws UnavailableException {
+    // JournalContext is closed before block deletion context so that file system master changes
+    // get written before block master changes. If a failure occurs between deleting an inode and
+    // remove its blocks, it's better to have an orphaned block than an inode with a missing block.
+    closeQuietly(mJournalContext);
+    closeQuietly(mBlockDeletionContext);
+
+    if (mThrown != null) {
+      Throwables.propagateIfPossible(mThrown, UnavailableException.class);
+      throw new RuntimeException(mThrown);
+    }
+  }
+
+  private void closeQuietly(AutoCloseable c) {
```

Comment:
isn't this what Guava closer has done for us?

Reply:
It's similar, but we need guarantees about the order of closing, and Guava `Closer` does not provide them https://google.github.io/guava/releases/19.0/api/docs/com/google/common/io/Closer.html

I'll add a comment to the class javadoc to mention this

Reply:
Just FYI, I believe Guava closer is closing registered objects in a LIFO order
https://github.com/google/guava/blob/master/guava/src/com/google/common/io/Closer.java#L210
I will leave this up to you whether to leverage Guava, if LIFO is sufficient